Partner Portal: Add the 'credit-card' redux store

IanRamosC opened the pull request:
#### Changes proposed in this Pull Request

This PR adds the base store for the 'credit-card' payment method store that is going to be used in the Partner Portal. It's pretty much a port from what's currently used in the composite-checkout package and might need improvements later.

This is going to be used behind a feature flag, so it won't harm having this merged

This is part of #53795

#### Testing instructions
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Unit tests have been added for actions and selectors

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #53795

IanRamosC committed:
Partner Portal: Add the 'credit-card' redux store

This PR adds the base store for the 'credit-card' payment method store that is going to be used in the Partner Portal. It's pretty much a port from what's currently used in the composite-checkout package and might need improvements later.

This is going to be used behind a feature flag, so it won't harm having this merged

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/hooks/use-create-stored-credit-card.ts:
```diff
@@ -0,0 +1,44 @@
+} ): PaymentMethod | null {
+	const shouldLoadStripeMethod = ! isStripeLoading && ! stripeLoadingError;
+	const stripePaymentMethodStore = useMemo( () => createStoredCreditCardPaymentMethodStore(), [] );
+	const stripeMethod = useMemo(
```
This (`stripeMethod`) could be consider an unnecessary `const` since we just declare it and return it

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/stored-credit-card-method.js:
```diff
@@ -0,0 +1,83 @@
+	);
+}
+
+function CreditCardLogos() {
```
Some of the functions we define here like `CreditCardLogos` or `CreditCardLabel` are 100% duplicates of the code from Blue/my-sites.

I don't need this to be changed for us to merge this PR since we might want to move faster and independent from Blue but we already use components from the composite checkout library like `<PaymentLogo>` which makes us dependent on the origin of this code.
We generally only depend on imports from `@automattic/...` a couple of places so I don't think we're THAT dependent on the origin of this code that we need to abstract logic to make them commonly accessible but in theory it could be DRY'er for the entire repository if we could import them here and in `client/my-sites/checkout/composite-checkout/payment-methods/credit-card/index.js` instead of defining them twice.

IanRamosC replied:
We can actually remove this entirely since it's not being used in our codebase. Thanks for pointing that out!

kallehauge commented on client/state/partner-portal/payment-methods/actions.ts:
```diff
@@ -0,0 +1,28 @@
+/**
+ * External dependencies
+ */
+import { AnyAction } from 'redux';
```
What are the reasonsbehind defining the actions here compared to getting them from `createCreditCardPaymentMethodStore` (`client/my-sites/checkout/composite-checkout/payment-methods/credit-card/index.js`) ?

I'm asking just as much our of curiosity to get more familiar with React, stores and redux 😄 
Is it to avoid loading the entire store? And if so, could we abstract them out and make them importable separately so we don't have to maintain these 1:1 to extend the Blue checkout logic? Or would it just make more sense for us to keep them as is? 🧠 
If it's strictly to make them more easily testable, would it make sense to add the tests to the origin instead?

IanRamosC replied:
The main reason for having a port instead of improving the existing one is that we would have to increase the complexity of the current composite-checkout logic to address both scenarios and them both are quite different. So instead of creating a huge one that is going to be loaded when not necessary to cover 2 not-so-equal scenarios, we decided to have separate ones that could be extended on their own if needed.

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/stored-payment-method-api.ts:
```diff
@@ -0,0 +1,81 @@
+	cardToken: string;
+	stripeConfiguration: StripeConfiguration;
+	purchase?: Purchase | undefined;
+	useAsPrimaryPaymentMethod?: boolean;
```
I would love to get this refactored to not be a three-state boolean so we don't rely on inline comments like `// if undefined, we do not add this property` for this to make sense and why wouldn't we want `use_for_existing` to be set for all requests to begin with? 😅 
Just a comment. Not something we should act on now 😄

kallehauge reviewed (commented):
**Purchases**

After looking at [`assignNewCardProcessor` in the big PR](https://github.com/Automattic/wp-calypso/pull/53795/files#diff-fec6808b55acdc1d13b866b1c1f2ba1cb4a282251d14ac9f0b6f2375be9c3d9bR64) (#53795) then it does looks like purchases are optional and could be rendered void for us and therefore be removed completely from our logic which would simplify it even further.
But this does sound like something that requires further digging (potentially after we have added context to the API endpoints) to be entirely sure. So something to keep in mind for the near future!

**General comment**

I've started to confuse myself a bit more than might be necessary because we depend on the same endpoints as my-sites/blue but we make our own implementation. How much of this do we actually have to implement as a port or could we expand the existing in-production version since we are going to share part of the same end-logic?

I realised this after I started making question about abstracting out some code because it's identical. I am, however, perfectly fine with us continuing with our own version that might look a bit duplicate sometimes just so we can move faster and become smarter and then have a separate project in the future where we share our new knowledge and try and move our work into the origin and maybe bypass some of the future maintenance required

IanRamosC committed:
Partner Portal: Remove unused code

IanRamosC committed:
Partner Portal: Remove unused code

IanRamosC committed:
Partner Portal: Add initial state for payment method

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/get-payment-method-id-from-payment.ts:
```diff
@@ -0,0 +1,14 @@
+		return 'existingCard-' + payment.creditCard.id;
+	}
+	return 'none';
+}
```
I don't think we ever use this function (`getPaymentMethodIdFromPayment`) and was made irrelevant after the latest changes where we removed purchases - I've tried to find a use for it in PR #53795 as well but I cannot find anywhere it's being imported.

The only case I can think of is if we plan to use this when updating cards and we create a purchase under the hood that has data we depend on during the update action

IanRamosC replied:
Thanks, this was a leftover indeed.

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/hooks/use-create-stored-credit-card.ts:
```diff
@@ -0,0 +1,44 @@
+import { createStoredCreditCardMethod } from 'calypso/jetpack-cloud/sections/partner-portal/payment-methods/stored-credit-card-method';
+import { createStoredCreditCardPaymentMethodStore } from 'calypso/state/partner-portal/payment-methods/';
+
+export function useCreateStoredCreditCard( {
```
Would it make sense to rename this to `useCreateStoredCreditCardMethod` instead since we specifically try to return a `createStoredCreditCardMethod` here and not `createStoredCreditCard` ? 😄

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/stored-credit-card-method.js:
```diff
@@ -0,0 +1,33 @@
+				activeButtonText={ activePayButtonText }
+			/>
+		),
+		getAriaLabel: ( __ ) => __( 'Credit Card' ),
```
Question! We import `useTranslate` from `i18n-calypso` in `assignNewCardProcessor` when we want to declare a string as translatable.

Is there some magic behind `( __ ) => __( 'Credit Card' )` that I'm not aware of or is this just an oversight? (I cannot tell you specifically if this is an oversight or not, hence the question 😅 )

IanRamosC replied:
The component that uses the `getAriaLabel` calls it passing down the `localize` as the [argument to that function](https://github.com/Automattic/wp-calypso/tree/trunk/packages/composite-checkout#payment-methods). So, in this case, `__` is just the argument name for the `localize` function.

kallehauge commented on client/state/partner-portal/payment-methods/reducer.ts:
```diff
@@ -0,0 +1,105 @@
+import { combineReducers } from 'calypso/state/utils';
+import { maskField } from 'calypso/lib/checkout';
+
+const initialState = {
```
🚨  **_Warning: Veeeery nitpicky and opinionated!_** 🚨 

Technically this contains multiple different initial states (plural) and not one specific initial state for e.g. brand, so `initialState` => `initialStates` would improve MY readability of the code.
I feel like it should objectively make the code more readable as well but it's very much mentioned with my own experience in mind and the this comment can easily be skipped.

IanRamosC replied:
At the end of the file, we use `combineReducers` which will merge those reducers into only one. The `initialState` is usually used to define the default state of a reducer.

kallehauge commented on client/state/partner-portal/payment-methods/reducer.ts:
```diff
@@ -0,0 +1,105 @@
+				...state,
+				[ action.payload.key ]: {
+					value: maskField( action.payload.key, state[ action.payload.key ], action.payload.value ),
+					isTouched: true,
```
💡  **Thought - this isn't a required action** 💡 

I noticed that I specifically had to look up what `isTouched` does whereas imported functions like `maskField` make sense because they have one place of origin where you can look them up and my editor is also able to grab the function documentation and easily give me a quick overview of what is happening.

`isTouched` doesn't have a single place of origin/documentation but rather exists as part of multiple functions in e.g. `client/my-sites/checkout/composite-checkout/types/wpcom-store-state.ts` with assumed knowledge of context.
To make it easier for onboarding Infinity developers then this might need an inline comment saying "Mark field as touched so..." to eliminate assumptions or required digging for context?

kallehauge commented on client/state/partner-portal/payment-methods/selectors.ts:
```diff
@@ -0,0 +1,22 @@
+
+export function getFields( state ) {
+	return state.fields;
+}
```
Would it make sense to add a type declaration to the selector functions?

kallehauge reviewed (commented):
You've done an awesome job reducing this PR to a slimmed down version @IanRamosC ! 💪 

The size and general state of the code is much easier to review and get an overview of now. Looking at the type of feedback I've left as part of this review then it also indicated that we're standing right in front of the goal line!

Half of my comments are just that. Comments/questions and not necessarily requests for changes. After these have been address then I don't anticipate I'll have any further comments 😄

IanRamosC committed:
Parner Portal: Address latest feedback

This commit removes unnecessary code, and improve the type definitions and readability for the `payment-methods` store.

kallehauge commented on client/state/partner-portal/payment-methods/reducer.ts:
```diff
@@ -0,0 +1,105 @@
+				...state,
+				[ action.payload.key ]: {
+					value: maskField( action.payload.key, state[ action.payload.key ], action.payload.value ),
+					isTouched: true, // mark whether the field has been touched
```
My idea with this comment was more "What does it mean when something has been touched?" Like:
`isTouched means that the field have a changed value or has been in a "focus" event state and requires new validation` or something like that (I'm not 100% sure that is actually what happens 😂 ).

The current version doesn't really provide much from a context perspective to enlighten an onboarding developer of what this actually means that we set the `isTouched` state to true 😄

kallehauge reviewed (approved):
@IanRamosC I just have the one comment about the `isTouched` documentation, then it all LGTM 👍

IanRamosC committed:
Partner Portal: Add more context to isTouched description comment

sirbrillig commented on client/jetpack-cloud/sections/partner-portal/payment-methods/stored-credit-card-method.js:
```diff
@@ -0,0 +1,33 @@
+ * Internal dependencies
+ */
+import CreditCardFields from 'calypso/jetpack-cloud/sections/partner-portal/credit-card-fields';
+import CreditCardPayButton from 'calypso/jetpack-cloud/sections/partner-portal/credit-card-fields/credit-card-pay-button';
```
👋  Hi! This imports a file that doesn't exist. Was that intentional as part of an ongoing project, or is this abandoned? 

The reason I ask is that, since this file isn't TS, it's causing a type error in `client/jetpack-cloud/sections/partner-portal/payment-methods/hooks/use-create-stored-credit-card.ts` because it mistakenly thinks that the `activePayButtonText` prop must be undefined. In trying to fix that type error, I began converting this file to TS only to discover that it is invalid.

sirbrillig replied:
Stripping out some of the non-existent imports in https://github.com/Automattic/wp-calypso/pull/59545

IanRamosC replied:
Hi @sirbrillig. Kinda missed the ping somehow, but thank you for stripping out the unused code. That code was part of a major feature team Infinity was working on by that time, but we had to focus on other high priority projects.
cc @Automattic/jetpack-infinity

sirbrillig replied:
No problem! Feel free to re-add whatever code is needed when you get around to it!
